Add support for shard latency checking

Brayzure opened the pull request:
- Each shard now stores the last time it sent a heartbeat, and when it was last acknowledged
- Shard getter returns latency if both values exist

Brayzure committed:
Add support for shard latency checking

- Each shard now stores the last time it sent a heartbeat, and when it was last acknowledged
- Shard getter returns latency if both values exist

hsiW reviewed (changes requested):
Forgot to add docs for new things

Brayzure committed:
Add documentation
- Oops, forgot

abalabahaha commented on lib/gateway/Shard.js:
```diff
@@ -42,6 +45,10 @@ class Shard extends EventEmitter {
     }
 
+    get latency() {
+        return this.lastHeartbeatSent && this.lastHeartbeatReceived ? this.lastHeartbeatReceived - this.lastHeartbeatSent : null;
```
`null` should probably be a JS number. Maybe `Infinity`?

Brayzure replied:
`Infinity` would certainly work, but I'm wondering if there's a better way to signify that the value the user is asking for doesn't make sense. Perhaps `NaN`?

abalabahaha commented on lib/gateway/Shard.js:
```diff
@@ -31,6 +31,9 @@ var Inflator = typeof window !== "undefined" && Pako ? Pako.inflate : Zlib.infla
 * @prop {Array<String>?} discordServerTrace Debug trace of Discord servers
 * @prop {String} status The status of the shard. "disconnected"/"connecting"/"handshaking"/"connected"
+* @prop {Number} lastHeartbeatReceived Timestamp when Discord last acknowledged a heartbeat
+* @prop {Number} lastHeartbeatSent Timestamp when shard last sent a heartbeat
```
Wording is a bit awkward for these

Brayzure replied:
Yeah, I thought it was, but I couldn't figure out a better way to word it. Ideas?

Brayzure committed:
Requested changes
- Return Infinity if latency can't currently be calculated
- Clarify docs a bit
